runtime/service.go

```diff
@@ -1183,11 +1192,18 @@ func (s *service) shutdown(
 		s.shutdownLoop(requestCtx, timeout, req)
 	}()
 
-	err := s.machine.Wait(context.Background())
-	if err == nil {
-		return nil
+	var result *multierror.Error
+	if err := s.machine.Wait(context.Background()); err != nil {
+		result = multierror.Append(result, err)
+	}
+	if err := s.jailer.Close(); err != nil {
+		result = multierror.Append(result, err)
```

Comment:
Is this racing with monitorVmExit()'s call to jailer.Close() as well as some of the actions in shutdownLoop() which can exit the shim while this is in process?

Reply:
Good point. It might be. Let me double check.

Reply:
shutdownLoop() should be fine. Since `jailer.Close()` is always called after `machine.Wait()`, we can gurantee the goroutine is not there at the time we are calling jailer.Close.

Regarding monitorVMExit(), there is a chance that two jailer.Close() in happening in the same time. I'm not so sure how bad it would be.

I do agree our shutdown logic is complicated.

Reply:
How about putting jailer.Close() and publishVMStop() inside sync.Once and put the once inside service as like vmStartOnce?

Reply:
I think moving all the cleanup stuff to one func() and putting that inside a sync.Once could make sense. We'd need a way to also show when that function is done that it's safe to call shimCancel I think.